test/common/http/async_client_impl_test.cc

```diff
@@ -701,6 +703,64 @@ TEST_F(AsyncClientImplTest, WithMetadata) {
   response_decoder_->decodeData(data, true);
 }
 
+class TestStateObject : public StreamInfo::FilterState::Object {
+public:
+  TestStateObject(std::string value) : value_(value) {}
+
+  const std::string& value() const { return value_; }
+
+private:
+  std::string value_;
+};
+
+TEST_F(AsyncClientImplTest, WithFilterState) {
+  message_->body().add("test-body");
+  Buffer::Instance& data = message_->body();
+
+  EXPECT_CALL(cm_.thread_local_cluster_.conn_pool_, newStream(_, _, _))
+      .WillOnce(Invoke(
+          [&](ResponseDecoder& decoder, ConnectionPool::Callbacks& callbacks,
+              const ConnectionPool::Instance::StreamOptions&) -> ConnectionPool::Cancellable* {
+            callbacks.onPoolReady(stream_encoder_, cm_.thread_local_cluster_.conn_pool_.host_,
+                                  stream_info_, {});
+            response_decoder_ = &decoder;
+            return nullptr;
+          }));
+
+  EXPECT_CALL(cm_.thread_local_cluster_, httpConnPool(_, _, _))
+      .WillOnce(Invoke([&](Upstream::ResourcePriority, absl::optional<Http::Protocol>,
+                           Upstream::LoadBalancerContext* context) {
+        const StreamInfo::FilterState& filter_state = context->requestStreamInfo()->filterState();
+        const TestStateObject* state = filter_state.getDataReadOnly<TestStateObject>("test-filter");
+        EXPECT_NE(state, nullptr);
+        EXPECT_EQ(state->value(), "stored-test-state");
+        return Upstream::HttpPoolData([]() {}, &cm_.thread_local_cluster_.conn_pool_);
+      }));
+
+  TestRequestHeaderMapImpl copy(message_->headers());
+  copy.addCopy("x-envoy-internal", "true");
+  copy.addCopy("x-forwarded-for", "127.0.0.1");
+  copy.addCopy(":scheme", "http");
+
+  EXPECT_CALL(stream_encoder_, encodeHeaders(HeaderMapEqualRef(&copy), false));
+
+  AsyncClient::RequestOptions options;
+  auto state_object = std::make_shared<TestStateObject>("stored-test-state");
+  auto filter_state =
+      std::make_shared<StreamInfo::FilterStateImpl>(StreamInfo::FilterState::LifeSpan::FilterChain);
+  filter_state->setData("test-filter", state_object, StreamInfo::FilterState::StateType::Mutable);
+  options.setFilterState(filter_state);
```

Comment:
Is the idea that a client must explicitly set the filter state in-code? Or would it be something automatic for sidecalls like ext_authz?